src/main/flight/rpm_filter.c

```diff
@@ -45,200 +45,155 @@
+FAST_DATA_ZERO_INIT static float erpmToHz;
 
-FAST_DATA_ZERO_INIT static uint8_t currentMotor;
-FAST_DATA_ZERO_INIT static uint8_t currentHarmonic;
-FAST_DATA_ZERO_INIT static uint8_t currentFilterNumber;
-FAST_DATA static rpmNotchFilter_t *currentFilter = &filters[0];
+// batch processing of RPM notches
+FAST_DATA_ZERO_INIT static int notchUpdatesPerIteration;
+FAST_DATA_ZERO_INIT static int motorIndex;
+FAST_DATA_ZERO_INIT static int harmonicIndex;
 
 
-
-PG_REGISTER_WITH_RESET_FN(rpmFilterConfig_t, rpmFilterConfig, PG_RPM_FILTER_CONFIG, 5);
-
-void pgResetFn_rpmFilterConfig(rpmFilterConfig_t *config)
+void rpmFilterInit(const rpmFilterConfig_t *config, const timeUs_t looptimeUs)
 {
-    config->rpm_filter_harmonics = 3;
-    config->rpm_filter_min_hz = 100;
-    config->rpm_filter_fade_range_hz = 50;
-    config->rpm_filter_q = 500;
-
-    config->rpm_filter_lpf_hz = 150;
-}
-
-static void rpmNotchFilterInit(rpmNotchFilter_t *filter, const rpmFilterConfig_t *config, const timeUs_t looptimeUs)
-{
-    filter->harmonics = config->rpm_filter_harmonics;
-    filter->minHz = config->rpm_filter_min_hz;
-    filter->maxHz = 0.48f * 1e6f / looptimeUs; // don't go quite to nyquist to avoid oscillations
-    filter->fadeRangeHz = config->rpm_filter_fade_range_hz;
-    filter->q = config->rpm_filter_q / 100.0f;
-    filter->looptimeUs = looptimeUs;
-
-    for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
-        for (int motor = 0; motor < getMotorCount(); motor++) {
-            for (int i = 0; i < filter->harmonics; i++) {
-                biquadFilterInit(
-                    &filter->notch[axis][motor][i], filter->minHz * i, filter->looptimeUs, filter->q, FILTER_NOTCH, 0.0f);
-            }
-        }
-    }
-}
+    motorIndex = 0;
+    harmonicIndex = 0;
+    minMotorFrequencyHz = 0;
+    rpmFilter.numHarmonics = 0; // disable RPM Filtering
 
-void rpmFilterInit(const rpmFilterConfig_t *config)
-{
-    currentFilter = &filters[0];
-    currentMotor = currentHarmonic = currentFilterNumber = 0;
-
-    numberRpmNotchFilters = 0;
+    // if bidirectional DShot is not available
     if (!motorConfig()->dev.useDshotTelemetry) {
-        gyroFilter = NULL;
         return;
     }
 
-    pidLooptime = gyro.targetLooptime;
-    if (config->rpm_filter_harmonics) {
-        gyroFilter = &filters[numberRpmNotchFilters++];
-        rpmNotchFilterInit(gyroFilter, config, pidLooptime);
-    } else {
-        gyroFilter = NULL;
-    }
-
+    // init LPFs for RPM data
     for (int i = 0; i < getMotorCount(); i++) {
-        pt1FilterInit(&rpmFilters[i], pt1FilterGain(config->rpm_filter_lpf_hz, pidLooptime * 1e-6f));
+        pt1FilterInit(&motorFreqLpf[i], pt1FilterGain(config->rpm_filter_lpf_hz, looptimeUs * 1e-6f));
     }
 
-    erpmToHz = ERPM_PER_LSB / SECONDS_PER_MINUTE  / (motorConfig()->motorPoleCount / 2.0f);
+    // if RPM Filtering is configured to be off
+    if (!config->rpm_filter_harmonics) {
+        return;
+    }
```

Comment:
No, that's intended because of historical reasons. Getting RPM telemetry was (and still is) part of the RPM filter. To make use of RPM data while RPM Filtering is disabled we need this split up. So first you need to check if bidir Dshot is active, then you initialize the RPM data filters, then you check if RPM Filter is active, and if so you initialize the RPM Filter.

I am planning to refactor the Dshot part out of the RPM Filter in BF 4.5 so this doesn't look weird anymore.

Reply:
Yes was thinking it was doing something in a place it was not supposed to be. Thanks for the heads up.

Reply:
Even if the RPM filtering is off, the RPM data is still used in logs and for dynamic idle.  I'm sure it can be tidied up.